crates/engine/tree/src/tree/instrumented_state.rs

```diff
@@ -0,0 +1,304 @@
+//! Implements a state provider that tracks latency metrics.
+use alloy_primitives::{Address, StorageKey, StorageValue, B256};
+use metrics::{Gauge, Histogram};
+use reth_errors::ProviderResult;
+use reth_metrics::Metrics;
+use reth_primitives_traits::{Account, Bytecode};
+use reth_provider::{
+    AccountReader, BlockHashReader, HashedPostStateProvider, StateProofProvider, StateProvider,
+    StateRootProvider, StorageRootProvider,
+};
+use reth_trie::{
+    updates::TrieUpdates, AccountProof, HashedPostState, HashedStorage, MultiProof,
+    MultiProofTargets, StorageMultiProof, StorageProof, TrieInput,
+};
+use std::{
+    sync::atomic::{AtomicU32, AtomicU64, Ordering},
+    time::{Duration, Instant},
+};
+
+/// Nanoseconds per second
+const NANOS_PER_SEC: u32 = 1_000_000_000;
+
+/// An atomic version of [`Duration`], using an [`AtomicU64`] to store the seconds part of the
+/// duration, and an [`AtomicU32`] for the nanoseconds part.
+#[derive(Default)]
+pub(crate) struct AtomicDuration {
+    /// The seconds part of the duration
+    seconds: AtomicU64,
+
+    /// The nanoseconds part of the duration
+    nanos: AtomicU32,
+}
+
+impl AtomicDuration {
+    /// Returns a zero duration.
+    pub(crate) const fn zero() -> Self {
+        Self { seconds: AtomicU64::new(0), nanos: AtomicU32::new(0) }
+    }
+
+    /// Returns the duration as a [`Duration`]
+    pub(crate) fn duration(&self) -> Duration {
+        Duration::new(self.seconds.load(Ordering::Relaxed), self.nanos.load(Ordering::Relaxed))
+    }
+
+    /// Adds a [`Duration`] to the atomic duration.
+    pub(crate) fn add_duration(&self, duration: Duration) {
+        // add the seconds part of the duration
+        self.seconds.fetch_add(duration.as_secs(), Ordering::Relaxed);
+        // add the nanoseconds part of the duration
+        self.nanos.fetch_add(duration.subsec_nanos(), Ordering::Relaxed);
+
+        // carry over the nanoseconds to the seconds part if the nanoseconds part is greater than 1
+        // second
+        if self.nanos.load(Ordering::Relaxed) >= NANOS_PER_SEC {
+            // get the number of seconds to carry over
+            let carry_over = self.nanos.load(Ordering::Relaxed) / NANOS_PER_SEC;
+            // set the nanoseconds part to the remainder
+            self.nanos.fetch_sub(carry_over * NANOS_PER_SEC, Ordering::Relaxed);
+            // add the carry over to the seconds part
+            self.seconds.fetch_add(carry_over as u64, Ordering::Relaxed);
+        }
+    }
```

Comment:
The only thing left in this PR is to improve this block. If we don't expect to accumulate `nanos` to the limit we can remove this. This mainly uses atomics because the state provider needs to be `Send + Sync` (so we can't use `RefCell` etc) and we need to be able to track cumulative time metrics with only a `&self` reference. The only solutions I can think of here are:
* Relax the `Sync` bound on state provider. we could then use refcell - we only expect this to be used in a single thread
* Atomics
* Mutex

Reply:
just switched to atomic u64